client/test/multiledger_dispute.go

```diff
@@ -0,0 +1,124 @@
+	"perun.network/go-perun/wire"
+)
+
+// TestMultiLedgerDispute runs an end-to-end test of the multi-ledger
+// functionality in the dispute case for the implementation specified in the
+// test setup.
+//nolint:revive // test.Test... stutters but this is OK in this special case.
+func TestMultiLedgerDispute(t *testing.T, mlt MultiLedgerSetup, challengeDuration uint64) {
+	ctx := context.Background()
+	require := require.New(t)
+	alice, bob := mlt.Client1, mlt.Client2
+
+	// Define initial balances.
+	//nolint:gomnd // We allow the balances to be magic numbers.
+	initBals := channel.Balances{
+		{big.NewInt(10), big.NewInt(0)}, // Asset 1.
+		{big.NewInt(0), big.NewInt(10)}, // Asset 2.
+	}
+	//nolint:gomnd
+	updateBals1 := channel.Balances{
+		{big.NewInt(5), big.NewInt(5)}, // Asset 1.
+		{big.NewInt(3), big.NewInt(7)}, // Asset 2. //nolint:mnd
+	}
+
+	// Establish ledger channel between Alice and Bob.
+
+	// Create channel proposal.
+	parts := []wire.Address{alice.WireAddress, bob.WireAddress}
+	initAlloc := channel.NewAllocation(len(parts), mlt.Asset1, mlt.Asset2)
+	initAlloc.Balances = initBals
+	prop, err := client.NewLedgerChannelProposal(
+		challengeDuration,
+		alice.WalletAddress,
+		initAlloc,
+		parts,
+	)
+	require.NoError(err, "creating ledger channel proposal")
+
+	// Setup proposal handler.
+	channels := make(chan *client.Channel, 1)
+	errs := make(chan error)
+	go alice.Handle(
+		AlwaysRejectChannelHandler(ctx, errs),
+		AlwaysAcceptUpdateHandler(ctx, errs),
+	)
+	go bob.Handle(
+		AlwaysAcceptChannelHandler(ctx, bob.WalletAddress, channels, errs),
+		AlwaysAcceptUpdateHandler(ctx, errs),
+	)
+
+	// Open channel.
+	chAliceBob, err := alice.ProposeChannel(ctx, prop)
+	require.NoError(err, "opening channel between Alice and Ingrid")
+	var chBobAlice *client.Channel
+	select {
+	case chBobAlice = <-channels:
+	case err := <-errs:
+		t.Fatalf("Error in go-routine: %v", err)
+	}
+
+	// Start Bob's watcher.
+	go func() {
+		errs <- chBobAlice.Watch(bob)
+	}()
+
+	// Notify Bob when an update is complete.
+	done := make(chan struct{}, 1)
+	chBobAlice.OnUpdate(func(from, to *channel.State) {
+		done <- struct{}{}
+	})
+
+	// Update channel.
+	err = chAliceBob.Update(ctx, func(s *channel.State) {
+		s.Balances = updateBals1
+	})
+	require.NoError(err)
+
+	// Wait until Bob's watcher processed the update.
+	<-done
+	time.Sleep(100 * time.Millisecond) //nolint:gomnd // The 100ms is a guess on how long the watcher needs to catch up.
+
+	// Store state.
+	req1 := client.NewTestChannel(chAliceBob).AdjudicatorReq()
+
+	// Alice registers state on l1 adjudicator.
+	err = mlt.Adjudicator1.Register(ctx, req1, nil)
+	require.NoError(err)
+
+	e := <-bob.Events
+	require.IsType(e, &channel.RegisteredEvent{})
+
+	// Close channel.
+	err = chBobAlice.Settle(ctx, false)
+	require.NoError(err)
+
+	// Check final balances.
+	require.True(mlt.BalanceReader1.Balance(alice.WalletAddress, mlt.Asset1).Cmp(updateBals1[0][1]) == 0)
```

Comment:
This balance checking does not work for other backends where we cannot assume that the account balances were 0 and now only have the final balance of the channel. Also, there may be fees that we also would have to consider.
If we want to keep the balance checking in the tests we may have to make it more abstract.

Maybe we could add `type AssertBalance func(a addr, asset Asset, finalBal *big.Int)` as a parameter.

Reply:
Addressed by revising the balance checking in the client tests and the funding and withdrawal functionality of the mock backend.